### DISCUSSION TITLE
K8s operator - docs questions.

### DISCUSSION THREAD

**Discussion by PaulRudin** [v6d-io/v6d#1962] [General]
I'm looking through the docs for the vineyard k8s operator; and there are a couple of things that I don't fully understand. https://v6d.io/notes/cloud-native/vineyard-operator.html

In https://v6d.io/notes/cloud-native/vineyard-operator.html#create-a-vineyard-cluster, it mentions a deployment for 3-replica vineyard servers, but there's no explicit mention of 3 in the example manifest. Is 3 a default?

Regarding the GlobalObject and LocalObject CRDs. Presumably it's normal for vineyard objects to be created by application code; are we saying that when this happens something dynamically creates the corresponding k8s resources?

Regarding the scheduler. There's discussion of "jobs", but the examples show the creation of deployments, so these are not "jobs" in the k8s sense. Normally we think of k8s deployments as long lived; whereas jobs perform a single task and then end. The docs refer to https://github.com/v6d-io/v6d/tree/main/k8s/test/e2e/workflow-demo, and the manifests in this example do indeed create k8s jobs, not deployments.

**Comment by PaulRudin** [comment]
In the assembly section: what's the mechanism by which the scheduler knows when `assembly-job2` can be run? There's clearly an intended sequencing of assembly-job1 followed by assembly-job2, but I don't see what it is that defines this intention.

**Reply by dashanji** [reply]
The annotation `scheduling.k8s.v6d.io/required: assembly-job1` defined in `assembly-job2` determines the execution order. Also, the label `assembly.v6d.io/enabled: "true"` indicates the `assembly-job2` need to trigger an assembly operation, the scheduler will pend it until the assembly operation done. In a word, the execution order is `assembly-job1` -> `assembly job` -> `assembly-job2`

**Reply by PaulRudin** [reply]
Ah, OK I see, so you can in fact create a chain of dependencies. Or, since you can specify more multiple jobs, you can create a could create a DAG of dependencies (or a cycle which presumably would break things).

This is sort of like kubeflow / airflow / whatever, but I guess they give you other things too.

**Reply by dashanji** [reply]
Yeah, here we just simulate a simple DAG.

**Comment by dashanji** [comment]
> In https://v6d.io/notes/cloud-native/vineyard-operator.html#create-a-vineyard-cluster, it mentions a deployment for 3-replica vineyard servers, but there's no explicit mention of 3 in the example manifest. Is 3 a default?

No, the 3 replicas is just an example, you can define any value.

> Regarding the GlobalObject and LocalObject CRDs. Presumably it's normal for vineyard objects to be created by application code; are we saying that when this happens something dynamically creates the corresponding k8s resources?

When you called the `persist` operation to make a globalObject or localObject persistent, the related k8s resources will be created.
For persistent object, you could refer to the https://v6d.io/notes/key-concepts/objects.html#transient-vs-persistent

> 
> Regarding the scheduler. There's discussion of "jobs", but the examples show the creation of deployments, so these are not "jobs" in the k8s sense. Normally we think of k8s deployments as long lived; whereas jobs perform a single task and then end. The docs refer to https://github.com/v6d-io/v6d/tree/main/k8s/test/e2e/workflow-demo, and the manifests in this example do indeed create k8s jobs, not deployments.

Thanks for pointing it out, it should be `jobs` rather than `deployments`. Sorry about the confusion, we'll fix it ASAP.

**Reply by PaulRudin** [reply]
The point about the 3 is a very minor one - the example yaml doesn't specify 3, but the text says that it creates 3 replicas.

**Reply by dashanji** [reply]
Maybe I confused the doc section. Whatever, you are right. We defined the 3 replicas as default.

**Comment by PaulRudin** [comment]
I experimented with the k8s operator before, but I ended up not using it. See e.g. https://github.com/v6d-io/v6d/discussions/1355.

I think I'll have another go, there are some things I want to do that it would probably help with.

**Reply by dashanji** [reply]
If you only want to share memory between containers from two pods, I think the blog (https://v6d.io/tutorials/kubernetes/data-sharing-with-vineyard-on-kubernetes.html) can help a lot.

If there are any questions, please feel free to raise here.
